servant-server/test/Servant/Server/RouterSpec.hs

```diff
@@ -0,0 +1,260 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeOperators #-}
+module Servant.Server.RouterSpec (spec) where
+
+import           Control.Monad                  (unless)
+import           Data.Proxy                     (Proxy(..))
+import           Data.Text                      (unpack)
+import           Network.HTTP.Types             (Status (..))
+import           Network.Wai                    (Application, responseBuilder)
+import           Network.Wai.Internal           (Response (ResponseBuilder))
+import           Test.Hspec
+import           Test.Hspec.Wai                 (get, shouldRespondWith, with)
+import           Servant.API
+import           Servant.Server
+import           Servant.Server.Internal
+
+spec = describe "Servant.Server.Internal.Router" $ do
+  routerSpec
+  distributivitySpec
+
+routerSpec :: Spec
+routerSpec = do
+  let app' :: Application
+      app' = toApplication $ runRouter router'
+
+      router', router :: Router
+      router' = tweakResponse (twk <$>) router
+      router = leafRouter $ \_ cont -> cont (Route $ responseBuilder (Status 201 "") [] "")
+
+      twk :: Response -> Response
+      twk (ResponseBuilder (Status i s) hs b) = ResponseBuilder (Status (i + 1) s) hs b
+      twk b = b
+
+  describe "tweakResponse" . with (return app') $ do
+    it "calls f on route result" $ do
+      get "" `shouldRespondWith` 202
+
+distributivitySpec :: Spec
+distributivitySpec =
+  describe "choice" $ do
+    it "distributes endpoints through static paths" $ do
+      endpoint1 `shouldHaveSameStructureAs` endpoint2
+    it "distributes nested routes through static paths" $ do
+      static1 `shouldHaveSameStructureAs` static2
+    it "distributes nested routes through dynamic paths" $ do
+      dynamic1 `shouldHaveSameStructureAs` dynamic2
+    it "properly reorders permuted static paths" $ do
+      permute1 `shouldHaveSameStructureAs` permute2
+    it "properly reorders permuted static paths in the presence of Raw in end" $ do
+      permute3 `shouldHaveSameStructureAs` permute4
+    it "properly reorders permuted static paths in the presence of Raw in beginning" $ do
+      permute5 `shouldHaveSameStructureAs` permute6
+    it "properly reorders permuted static paths in the presence of Raw in middle" $ do
+      permute7 `shouldHaveSameStructureAs` permute8
+    it "properly reorders permuted static paths in the presence of a root endpoint in end" $ do
+      permute9 `shouldHaveSameStructureAs` permute12
+    it "properly reorders permuted static paths in the presence of a root endpoint in beginning" $ do
+      permute10 `shouldHaveSameStructureAs` permute12
+    it "properly reorders permuted static paths in the presence of a root endpoint in middle" $ do
+      permute11 `shouldHaveSameStructureAs` permute12
+    it "properly handles mixing static paths at different levels" $ do
+      level3 `shouldHaveSameStructureAs` level4
```

Comment:
I wish this could be a bit easier to read and understand. Sadly, we don't have local type synonyms, so the tests and the types cannot be easily put close to each other.

I don't have a very good idea how to make this better, unfortunately.

Currently the following convention seems to be followed: ```someApi$N `shouldHaveSameStructureAs` someApi$(N + 1)``` for odd Ns. Maybe another convention would be easier. E.g. ```someApi `shouldHaveSameStructureAs` someApi'```. You only would have to come up with unique names for the different test cases.